core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -2788,6 +2788,18 @@ public static double log(BigDecimal d0, BigDecimal d1) {
     return Math.log(d0.doubleValue()) / Math.log(d1.doubleValue());
   }
 
+  /** SQL {@code LOG2(number)} function applied to double values. */
+  public static double log2(double d0) {
+    return (Double.isInfinite(log(d0, 2)) && d0 == 0.0) ? Double.NaN : log(d0, 2);
+  }
+
+  /** SQL {@code LOG2(number)} function applied to
+   * BigDecimal and double values. */
+  public static double log2(BigDecimal d0) {
+    return (Double.isInfinite(log(d0, 2)) && d0.doubleValue() == 0.0) ? Double.NaN : log(d0, 2);
```

Comment:
Is `NaN` definitely what we want to return?

Reply:
At present, we mainly cannot return correct results for log2(0), log2(0.0), etc. In mysql, these return null. I think this corresponds to NaN.

Reply:
if you are emulating MySQL you should return the exact same value in this case. Otherwise your code is wrong.

Reply:
So it should return NULL?

Reply:
What does MySql return? Does it return `Null`? `NaN` is not the same as `null`. As Mihai said: `if you are emulating MySQL you should return the exact same value in this case. Otherwise your code is wrong`.

Reply:
I try to change the result to null in sqlfunctions, but a null pointer exception will be reported when comparingResultSet is checked.

Reply:
comparingResultSet looks like it doesn't allow for nulls when iterating over the result set

Reply:
To a certain extent, I agree with your point of view, but if it is just for the case of log2(0)==null that a lot of calcite code needs to be changed, I am skeptical. I can even suggest that our handling of log2(0) should be stated in the document, and I feel that mysql handles this wrongly from the beginning. I studied presto and arrow-datafusion, and to be honest, they couldn't overcome this problem very well.

Reply:
@caicancai if you think MySQL is wrong you should file a bug with MySQL.

Reply:
@mihaibudiu Maybe you are right, I would rather try to fix calcite than modify mysql, but I am not sure whether to modify the comparingResultSet method, because obviously we found a lot of problems when adapting the log2 function, and the problems are getting more and more complex.
Thank you

Reply:
Of course Julian can give his opinion but I'm confused what the issue is. MySQL has some behavior and we want to mimic that behavior. What's keeping us from doing this?

Reply:
Maybe I'm not sure how to modify it correctly and how to modify it appropriately. I hope to adapt it with minimal changes. At the moment, I'm not sure whether it's appropriate to modify comparingResultSet. I will try to find the best solution.

Reply:
@tanclary To be honest, I could have completed the changes and handed them over to you for review, but during the adaptation process, I found that I had a lot of doubts about this, which led me to ask you many questions. Perhaps this is the reason why this PR has been discussed for so long. It is very Thank you for your patient answers.

Reply:
If you upload your latest commit, and tell me what test to run, I can help.

Reply:
Thank you very much, just need rust test

Reply:
![捕获](https://github.com/apache/calcite/assets/77189278/81255246-5c28-471a-8106-7367ef160b15)
@tanclary I'm not sure if this is a problem with calcite, but it seems like I have no idea about it.

Reply:
SAFE_* functions doesn't look like this will happen